tests/integration/pilot/trafficshifting/traffic_shifting_test.go

```diff
@@ -0,0 +1,194 @@
+	"istio.io/istio/pkg/test/framework/components/pilot"
+)
+
+//	Virtual service topology
+//
+//						 a
+//						|-------|
+//						| Host0 |
+//						|-------|
+//							|
+//							|
+//							|
+//		-------------------------------------
+//		|weight1	|weight2	|weight3	|weight4
+//		|b			|c			|d			|e
+//	|-------|	|-------|	|-------|	|-------|
+//	| Host0 |	| Host1	|	| Host2 |	| Host3 |
+//	|-------|	|-------|	|-------|	|-------|
+//
+//
+
+var (
+	ist   istio.Instance
+	hosts = []string{"b", "c", "d", "e"}
+	p     pilot.Instance
+	g     galley.Instance
+)
+
+const (
+	batchSize = 100
+
+	errorBand = 10.0
+
+	testDuration = 10 * time.Second
+
+	virtualService = `
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+ name: {{.Name}}
+ namespace: {{.Namespace}}
+spec:
+ hosts:
+ - {{.Host0}}
+ http:
+ - route:
+   - destination:
+       host: {{.Host0}}
+     weight: {{.Weight0}}
+   - destination:
+       host: {{.Host1}}
+     weight: {{.Weight1}}
+   - destination:
+       host: {{.Host2}}
+     weight: {{.Weight2}}
+   - destination:
+       host: {{.Host3}}
+     weight: {{.Weight3}}
+`
+)
+
+type VirtualServiceConfig struct {
+	Name      string
+	Host0     string
+	Host1     string
+	Host2     string
+	Host3     string
+	Namespace string
+	Weight0   int32
+	Weight1   int32
+	Weight2   int32
+	Weight3   int32
+}
+
+func TestMain(m *testing.M) {
+	framework.NewSuite("traffic_shifting", m).
+		RequireEnvironment(environment.Kube).
+		Label(label.CustomSetup).
+		SetupOnEnv(environment.Kube, istio.Setup(&ist, nil)).
+		Run()
+}
+
+func TestTrafficShifting(t *testing.T) {
+	// Traffic distribution
+	weights := map[string][]int32{
+		"0-100":       {0, 100},
+		"20-80":       {20, 80},
+		"50-50":       {50, 50},
+		"33-33-34":    {33, 33, 34},
+		"25-25-25-25": {25, 25, 25, 25},
+	}
+
+	framework.
+		NewTest(t).
+		RequiresEnvironment(environment.Kube).
+		Run(func(ctx framework.TestContext) {
+
+			g, _ = galley.New(ctx, galley.Config{})
+			p, _ = pilot.New(ctx, pilot.Config{Galley: g})
+
+			ns := namespace.NewOrFail(t, ctx, "traffic-shifting", true)
+
+			a := newEcho(t, ctx, ns, "a")
+			b := newEcho(t, ctx, ns, "b")
+			c := newEcho(t, ctx, ns, "c")
+			d := newEcho(t, ctx, ns, "d")
+			e := newEcho(t, ctx, ns, "e")
+			a.WaitUntilReadyOrFail(t, b, c, d, e)
+
+			for k, v := range weights {
+				t.Run(k, func(t *testing.T) {
+					v = append(v, make([]int32, 4-len(v))...)
+
+					vsc := VirtualServiceConfig{
+						"traffic-shifting-rule",
+						hosts[0],
+						hosts[1],
+						hosts[2],
+						hosts[3],
+						ns.Name(),
+						v[0],
+						v[1],
+						v[2],
+						v[3],
+					}
+
+					tmpl, _ := template.New("VirtualServiceConfig").Parse(virtualService)
+					var buf bytes.Buffer
+					tmpl.Execute(&buf, vsc)
+
+					g.ApplyConfigOrFail(t, ns, buf.String())
+
+					// TODO: Find a better way to wait for configuration propagation
+					time.Sleep(10 * time.Second)
```

Comment:
Echo now provides a query interface for the sidecar.  Could you use this here to query the Envoy config dump (using the `pkg/test/util/structpath` library for verification) instead of sleeping?

Reply:
I will replace sleep with query.